staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -224,30 +224,50 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 		ResourceVersion: resourceVersion,
 	}
 
-	// TODO: We should consider moving this lock below after the watchCacheEvent
-	// is created. In such situation, the only problematic scenario is Replace(
-	// happening after getting object from store and before acquiring a lock.
-	// Maybe introduce another lock for this purpose.
-	w.Lock()
-	defer w.Unlock()
-	previous, exists, err := w.store.Get(elem)
-	if err != nil {
+	if err := func() error {
+		// TODO: We should consider moving this lock below after the watchCacheEvent
+		// is created. In such situation, the only problematic scenario is Replace(
+		// happening after getting object from store and before acquiring a lock.
+		// Maybe introduce another lock for this purpose.
+		w.Lock()
+		defer w.Unlock()
+
+		previous, exists, err := w.store.Get(elem)
+		if err != nil {
+			return err
+		}
+		if exists {
+			previousElem := previous.(*storeElement)
+			watchCacheEvent.PrevObject = previousElem.Object
+			watchCacheEvent.PrevObjLabels = previousElem.Labels
+			watchCacheEvent.PrevObjFields = previousElem.Fields
+		}
+
+		w.updateCache(watchCacheEvent)
+		w.resourceVersion = resourceVersion
+		defer w.cond.Broadcast()
+
+		switch watchCacheEvent.Type {
+		case watch.Added:
+			return w.store.Add(elem)
+		case watch.Modified:
+			return w.store.Update(elem)
+		case watch.Deleted:
+			return w.store.Delete(elem)
+		default:
+			return fmt.Errorf("unexpected event type: %v", watchCacheEvent.Type)
+		}
+	}(); err != nil {
 		return err
 	}
-	if exists {
-		previousElem := previous.(*storeElement)
-		watchCacheEvent.PrevObject = previousElem.Object
-		watchCacheEvent.PrevObjLabels = previousElem.Labels
-		watchCacheEvent.PrevObjFields = previousElem.Fields
-	}
-	w.updateCache(watchCacheEvent)
-	w.resourceVersion = resourceVersion
 
+	// Avoid calling event handler under lock.
+	// This is safe as long as there is at most one call to processEvent in flight
+	// at any point in time.
 	if w.onEvent != nil {
 		w.onEvent(watchCacheEvent)
 	}
-	w.cond.Broadcast()
-	return updateFunc(elem)
```

Comment:
Should `updateFunc` be removed from the function parameters as well, since we're no longer calling it? Any callers that might be impacted by no longer calling the provided `updateFunc`?

Reply:
done